### ISSUE TITLE
NotNullIfNotNull should work on input parameters

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#66065]
**Version Used**: 75d5175ff602c3dc7c91732fc05606a8d19b4350

**Steps to Reproduce**:

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAOgBEoIBzGAewGcAXKAY2YoGFGEApgEEYEADYBPZlGYBuAgXQBmMphI8CAbwIldZFegAsJALIAKANoA5RqysBXMWICSAMxt3HYs6LADGrmYADhAAThBgmACUUQC6ZNhEAPwkIeFgcAnJqWER0Tp62vh6JSRQriTBuZEkAIQAvCQwXiQAZK056ST1jc1OUQWlpawAFqGMAO5NAlMAogAeHAJB7IwwZlHyxXoAvoMk+0am2Gbk2WkRmWedeQPbukVDuuYXGTeRm/slrxQAKowAZVYoVgdA2shIAHpISR7DABPMggIOKwBAgSBMwjBQeoAQAOABsREwIBIVAEoQErgpAhgSxIARIEFSLGkwEkTRalOplLpAgoCnuJD2+B2QA)

```cs
using System;
using System.Diagnostics.CodeAnalysis;

public class C
{
    public void M([NotNullIfNotNull(nameof(param2))] string? param, string? param2)
    {
        if (param2 != null && param == null)
            throw new Exception();
    }
    
    void M1(string? param, string param2)
    {
        M(param, param2);
        param.ToString(); // unexpected warning CS8602: Dereference of a possibly null reference.

    }
}
```

**Expected Behavior**: No warning is given

**Actual Behavior**: The indicated warning is given

**Remarks**: In general, nullability postconditions such as `[NotNull]` are allowed to be applied to non-`out` parameters. The use case for these is usually to write `Assert`-style methods. We should adjust the behavior of `[NotNullIfNotNull]` to work regardless of the ref kind of the "output" parameter indicated in the attribute argument.
